### ISSUE TITLE
Morrowind Support

### ISSUE THREAD

**Issue by Infernio** [wrye-bash/wrye-bash#479]
Let's start with this: This is a **lot** of work, and far off. Goal is implementing every feature that [Wrye Mash (Polemos' Fork)](https://github.com/polemion/Wrye-Mash-Polemos) has.
Basic support is implemented in [inf-morrowind](https://github.com/wrye-bash/wrye-bash/tree/inf-morrowind), enough to install mods, read header details about them, read and write BSAs, and to read save headers.
That branch currently has the problem that it causes Oblivion's BP to hang the game on loading, which is why it's not on nightly.

Detailed overview of everything that needs to be done:

## Installers Tab
Nothing to do (besides updating BAIN dirs and such), our Installers tab appears to be a strict superset of Wrye Mash's.

## Mods Tab
- [x] Subrecord headers use a uint32 instead of a uint16 to denote the size. Done in e59eae7d1f57a1048640bbcb847fc09f0bdc50b2, but I suspect this is the cause of the BP issue mentioned above.
- [ ] GRUPs do not exist. Instead of the TES4-style grouping:
```
GRUP
GMST
GMST
GMST
GRUP
GLOB
GLOB
```
Morrowind plugins simply list the records in a row, which creates an implicit grouping:
```
GMST
GMST
GMST
GLOB
GLOB
```
- [ ] Wrye Mash features to implement (or reject, if we don't want them):
  - [x] The usual File submenu
  - [ ] The Group submenu - has different default groups than what we have, so might be worth looking into that? If not, then this is done.
  - [x] The usual Rating submenu
  - [ ] Export
    - [ ] Dialogue
    - [ ] Scripts
  - [ ] Import
    - [ ] Dialogue
    - [ ] LCV Schedules
    - [ ] Merged Lists (dunno about this one, see Bashed Patch section below)
    - [ ] Scripts
    - [ ] Edit Replacers...
  - [ ] Clean with TES3cmd
  - [ ] Check with TES3lint
  - [x] Custom Commands - probably not, I have my own ideas for how to do something like this for all games
  - [x] Readme...
  - [x] Copy to esp/esm
  - [ ] Renumbers Refs
  - [ ] Repair Refs
  - [ ] Sort Records
  - [x] Statistics - our version is `Details`
  - [ ] Updaters...

### Load Order
- [x] LO is based on timestamps.
- [x] Active mods are stored in the INI - done in 9785e291c833c470c00a3264eac54d0702fadfe1.

### Bashed Patch
I have some ideas on how we could do this with no changes to the patch procedures, by (ab)using the dynamic typing of Python.
Will have to experiment and see. But regardless, we can't ship without:
 - [ ] An equivalent to `Mashed Lists`, i.e. at least the leveled list patcher must be available.

## INI Tweaks Tab
Wrye Mash has no such thing. It has INI Tweaks, but they're very poorly implemented via a context menu option in the Mods tab. No previews, no intuitive interface, files are loose - our tab is far superior here.

## Saves Tab
- [ ] Identical to mod files, but with a few record-specific differences and save-exclusive record types.
  - [x] Save header support implemented in b8b0e7aae742bfc658691ef55eb9734275731781. Not perfect, save headers don't seem to have player levels and time played.
- [ ] Wrye Mash features to implement:
  - [x] The usual File submenu
  - [ ] Remove
    - [ ] Debris Cells
    - [ ] Spawned Creatures
    - [ ] Edit Removers...
    - [ ] Edit Safe Cells...
  - [ ] Journal...
  - [x] Load Masters
  - [ ] Map Notes
  - [ ] Repair All
  - [ ] Statistics
  - [ ] Update Map

## Screenshots Tab
- [x] Next Shot setting must become game-specific - done in ec949a00f57d1808502bccfcade37ce8752e013a.

## BSAs Tab
- [x] No folder records here, only a long list of file records. Properly implemented in 6f9076fab901f99a03c5a0f4de5f4c39ac78190f.
- [ ] Wrye Mash has a small BSA Archives tab inside the mods tab, so to mirror Mash, this issue depends on #233.

## Utilities Tab
Again, I have my own ideas on how to implement this. Sometime into 308+, I'll open an issue to further investigate this - but till then, our version of Wrye Mash will simply be missing this.

## Other Things
Polemos' Fork has a very cool feature where it exposes the column context menu options above the tab via a wx.MenuBar:

![WryeMash_MenuBar](https://user-images.githubusercontent.com/2845240/69163574-bf86d400-0aee-11ea-8957-ef91b322a596.png)

A lot of users don't even know you can click on the column headers, so exposing the options like this is something I'd really like to add to WB as well. Blocked by #190, obviously.

**Comment by Metallicow** [comment]
Ah, about time.... Morro support! :)
[Morrowtality](https://youtu.be/BsQikNAKct0l)

**Comment by Metallicow** [comment]
Also in wx if you expose all the functions of the program in the App Menus with ids,
you can call those functions/methods from anywhere. This is a issue with the current state of wrye bash since it did not take this course initially.
You could for example add all features to a top level menu and if menu is visible, then you *could* hide it if you want to... The main reason for binding menu items is the keyboard shortcuts work from anywhere, tho with the current state you would have to check what keys are being pressed and analize whether you want to call whatever func/meth.

**Comment by Metallicow** [comment]
...So for example, the best thing to do is to make menu items for every normal menuitem in any menu and make an specific id for that, that way it is callable from anywhere. Then when id is called all you would have to do is check for wrye bash is in `Morro` mode and if it is call the code.
The biggest problem really is handling all the ids, and since wrye bash in its current state is way blown up with stuff everywhere, this makes the problem worse, since it didn't start with a MenuBar and ids in the first place.

**Comment by Metallicow** [comment]
The most simple example of how this should be implemented is
```python
mode = wryebashMode()
if mode == 'Morrowind':
    if getTabImOn() == 'Installers':
        callMorrowindFuncMeth(ID_NUMBER_BLABLABLA)
    elif anothertab....:
        ...
elif mode is 'whateverOtherGame':
   # do the code
```

**Comment by Infernio** [comment]
I found the cause of the BP issue that was mentioned here - I did a `zip(record.masters, record.master_sizes)`, but forgot to pad `master_sizes` before that. So `masters` had the correct paths in it, but `master_sizes` was empty. `zip()` worked as expected and returned nothing, so the BP ended up with no masters. Whoops.

**Comment by Utumno** [comment]
Posting here given the name of the branch - nice work on `inf-479-global-menu` @Infernio as usual

I had resisted this but, as Linus put it, trusting people means accepting their decisions, even if you disagree sometimes. And I do not even disagree for that matter, I always thought that the no menu was just peculiar Bash style (and some more screen estate), but hey it keeps confusing newcomers so to heck with it (plus you even added an option for old-timers).

Tiny notes `class OrderedDefaultDict(collections.OrderedDict, collections.defaultdict)` will blow on python 3 (OrderedDict implemented in C so incompatible bases), but then all dicts are ordered

**Comment by Infernio** [comment]
> Tiny notes class OrderedDefaultDict(collections.OrderedDict, collections.defaultdict) will blow on python 3 (OrderedDict implemented in C so incompatible bases), but then all dicts are ordered

Yup, this relies on us dropping it during py3, I'll add a `PY3:` note for it.

**Comment by leandor** [comment]
Global menu has an advantage: you can access it with Keyboard only by pressing Alt plus the accelerator keys! I personally like it :)

Also, great work on Morrowind support! It has all come full circle it seems! As Morrowind was where it all started :)

**Comment by Metallicow** [comment]
One thing to note about menus and IDs is that 0 should be avoided at all costs. This will cause errors on mac and it also isn't really a readable way to do an ID even if it is a temp popup type menu with items.

Also note: stuff like this does work in msw Ex: 'MyMenuItemSetting\tisOn' or 'MyMenuItemSetting\tisOff', but is far from any sort of standard and even if you have seen this in a menu item in some of my apps, it really should probably be avoided, as it accelerators at some point might be validated in some sort of way and this type of improv could become invalid at any point in the future lol.

**Comment by rtentser** [comment]
Any plans for openmw support?

**Comment by Infernio** [comment]
Morrowind first, we'll see about OpenMW then. But yeah, there's no reason why we couldn't support OpenMW at some point.
